### ISSUE TITLE
DNS Interface Behaviour on Cluster Fail

### ISSUE THREAD

**Issue by markbugejacba** [hashicorp/consul#1888]
On cluster failure, the Consul DNS interface remains active and responds to otherwise valid DNS requests with a `NOERROR` response code.

Is this by design? Does aspect of the DNS interface actually remain functional in this state? 

Please consider disabling the DNS interface on failure so that upstream components can reliably detect cluster partitions and/or failure.

**Comment by markbugejacba** [comment]
I've also noticed situations where a `SERVFAIL` is returned during an outage. Is there anything that can be derived by the scenarios represented by the two error codes?

**Comment by slackpad** [comment]
Related - https://github.com/hashicorp/consul/issues/1906.

**Comment by slackpad** [comment]
Hi @markbugejacba this is a bit of a subtle artifact of configuring Consul to allow stale DNS responses. Assuming that you've enabled https://www.consul.io/docs/agent/options.html#allow_stale, there will be a window after the outage occurs (outage being no Consul leader) where Consul servers will still provide DNS results based on their own in-memory store. Once the https://www.consul.io/docs/agent/options.html#max_stale window has been exceeded, it will attempt to retry the RPC requests internally, which will fail because there's no leader, so you will start getting `SERVFAIL` responses.

Does this explanation fit with how you've got things configured? This seems like the right behavior given that stale responses are allowed, and lets the cluster better ride out brief moments without a leader, such as during leader elections. We are also planning on making this more robust with some internal retries under https://github.com/hashicorp/consul/issues/1554.

/cc @sean-

**Comment by Oloremo** [comment]
Documentation about "allow_stale" is based on the idea to allow clients(not masters) to answer dns requests. So, I wonder, if it's true, what master will be act the same?

**Comment by slackpad** [comment]
Hi @Oloremo the `allow_stale` configuration is set on each client agent but is partly implemented on the Consul servers. Clients connect to any of the Consul servers, and if staleness is not allowed then read requests are internally forwarded to the leader. If staleness is allowed, then any server can fulfill the request, and the client decides if the result is fresh enough to use (otherwise it will re-request and turn off `allow_stale` just for that request). There's currently no caching in the client agents themselves.

**Comment by Oloremo** [comment]
@slackpad Let me rephrase this:
Currently all hosts in my network do resolv lookup via my consul *masters*. If I'll add allow_stale = true to *master* consul nodes, will this allow them to answer to dns loopup request for the "max_stale" period during the "no leader" situation? 

By dns loopup request I mean basic glibc one, like "host nodename.node.consul".

**Comment by slackpad** [comment]
@Oloremo ah ok - yes you'd need to set `allow_stale` on your Consul masters if they are handling DNS for your network. That should allow them to service requests in the absence of a leader, up to the maximum staleness you configure.

**Comment by Oloremo** [comment]
@slackpad Thank you!

**Comment by markbugejacba** [comment]
@slackpad We do indeed have `allow_stale` enabled on all of our servers, so that would explain the `NOERROR` responses.

Right now, I it appears that the DNS service remains active and will happily return data retrieved from other data centers (where the prefix of the other data center was explicitly specified,) whilst returning `SERVFAIL` only for calls that would otherwise have been serviced within the failed data center.

We're still keen to understand if it would be feasible to devise a method for DNS resolvers to reliably detect when the cluster in a given data center has failed, so as to avoid referring/forwarding DNS traffic to failed servers. Ideally, we'd like the option for the DNS interface to completely disable itself on any server that is no longer authoritative (ie. minority split brain or complete cluster failure.) for its local data.

**Comment by slackpad** [comment]
@markbugejacba if you are seeing a difference between remote and local behavior it sounds like there might be a difference in how `allow_stale` is configured on the agents. The `allow_stale` configuration of the agent _servicing the DNS request_ is what determines the behavior. If an agent (even if it's a server) in dc1 is handling a DNS request for a server in dc2, then it's the agent in dc1's `allow_stale` setting that gets passed along in the RPC request.

**Comment by foreversunyao** [comment]
@slackpad ,
      You said "There's currently no caching in the client agents themselves." , and is there any plan to add client cache function  in the future ? Because if network between consul-client and server, application can not resolve any dns, it is not a good situation

**Comment by slackpad** [comment]
> You said "There's currently no caching in the client agents themselves." , and is there any plan to add client cache function in the future ? Because if network between consul-client and server, application can not resolve any dns, it is not a good situation

We are currently looking into some ways to do this, but it's not yet committed to a release on our roadmap just yet.

**Comment by foreversunyao** [comment]
@slackpad thanks to reply.
